chain-patrol-ts/src/fetcher.ts

```diff
@@ -0,0 +1,92 @@
+import { LRUCache } from "lru-cache";
+import { createAssetListApiOptions, createAssetDetailsApiOptions } from "./utils";
+import { ApiOptions, AssetList, Asset, AssetDetails } from "./types";
+import { ASSET_TYPES } from "./constants";
+
+export default class AssetFetcher {
+  assetListUrl: string;
+  assetDetailsUrl: string;
+  assetStatus: string;
+  private apiKey: string;
+  private readonly MAX_TRIES: number;
+  private assetListCache: LRUCache<string, Asset[]>;
+  private assetDetailsCache: LRUCache<string, AssetDetails>;
+
+  constructor(apiKey: string, assetListUrl: string, assetDetailsUrl: string, assetStatus: string) {
+    this.MAX_TRIES = 3;
+    this.apiKey = apiKey;
+    this.assetListUrl = assetListUrl;
+    this.assetDetailsUrl = assetDetailsUrl;
+    this.assetStatus = assetStatus;
+    this.assetListCache = new LRUCache<string, Asset[]>({
+      max: 1000,
+    });
+    this.assetDetailsCache = new LRUCache<string, AssetDetails>({
+      max: 1000,
+    });
+  }
+
+  getAssetlist = async (endDate: string, startDate: string): Promise<Asset[] | undefined> => {
+    const key = `${endDate}-${startDate}`;
+    if (this.assetListCache.has(key)) {
+      return this.assetListCache.get(key);
+    }
+
+    let tries = 0;
+    while (tries < this.MAX_TRIES) {
+      try {
+        // Since we can only query the API
+        // with one `type` at a time
+        const apiOptions: ApiOptions[] = [];
+        ASSET_TYPES.map((type: string) => {
+          apiOptions.push(createAssetListApiOptions(this.apiKey, type, this.assetStatus, endDate, startDate));
+        });
```

Comment:
since there's no failure probability for these lines, these can be moved outside of the `while` block. otherwise these will be redundantly repeat themselves.